### PR TITLE
build: auto-generate the codesigning cert used for macOS CI testing runs

### LINKED ISSUE #17649: macOS autoUpdater spec fails on PRs from forks
### Preflight Checklist

* [x] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/master/CONTRIBUTING.md) for this project.
* [x] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
* [x] I have searched the issue tracker for an issue that matches the one I want to file, without success.

### Issue Details

* **Electron Version:**
  * Code from master
* **Operating System:**
  * macOS on Electron's CircleCI instances
* **Last Known Working Electron version:**:
  * Right before https://github.com/electron/electron/pull/17442 was merged

### Expected Behavior

The testsuite passes on CircleCI's `build-mac` workflow

### Actual Behavior

[1 test failure](https://circleci.com/gh/electron/electron/166983):

```
autoUpdater behavior "before each" hook for "should have a valid code signing identity" - "before each" hook for "should have a valid code signing identity"
/Users/distiller/project/src/electron/spec-main/api-autoupdater-darwin-spec.ts

Error: No valid signing identity available to run autoUpdater specs
    at Context.<anonymous> (electron/spec-main/api-autoupdater-darwin-spec.ts:26:15)
```

### To Reproduce

1. Fork `electron/electron` and make a branch
2. Upload branch to personal fork
3. Make a PR
4. Wait for CircleCI's `build-mac` workflow to run

### Screenshots

image

### Additional Information

CC: @MarshallOfSound as he told me to :smile:

### PR DESCRIPTION
**Pull request by MarshallOfSound**
Fixes #17649

Instead of using a pre-generated cert we can dynamically make a new random one every time.  Thanks to @nornagon for the optimism that this would be possible 😀 

Notes: no-notes

### CODE REVIEW

File: script/codesign/codesign.cnf [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+C = CA
+ST = BC
+L = Vancouver
+O = ElectronJS
+OU = BuildAutomation
+CN = codesign.electronjs.org
+
+[extended]
+keyUsage         = critical,digitalSignature
+extendedKeyUsage = critical,codeSigning
```
Comment (nornagon) [inline script/codesign/codesign.cnf]: is this extended usage actually required? the snippet you showed me didn't seem to have it

File: script/codesign/generate-identity.sh [hunk 2]
```diff
@@ -0,0 +1,32 @@
+#!/bin/sh
+
+dir="$(dirname $0)"/.working
+
+cleanup() {
+    mkdir -p "$dir"
+    rm -r "$dir"
```
Comment (nornagon) [inline script/codesign/generate-identity.sh]: no need for `mkdir` here, just `rm -rf`

File: script/codesign/generate-identity.sh [hunk 3]
```diff
@@ -0,0 +1,32 @@
+#!/bin/sh
+
```
Comment (nornagon) [inline script/codesign/generate-identity.sh]: set -eo pipefail

File: script/codesign/generate-identity.sh [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+security import "$dir"/private.pem -k $KEY_CHAIN
+
+# Generate Trust Settings
+node "$(dirname $0)"/gen-trust.js "$dir"/certificate.cer "$dir"/trust.xml
+
+# Import Trust Settings
+sudo security trust-settings-import -d "$dir/trust.xml"
+
+# Clean Up
+cleanup
```
Comment (nornagon) [inline script/codesign/generate-identity.sh]: perhaps `trap cleanup EXIT` at the top of the script?

per https://stackoverflow.com/a/2130323

File: script/codesign/codesign.cnf (hunk 1 above)
Comment (MarshallOfSound) [reply script/codesign/codesign.cnf]: @nornagon Yes, `security` doesn't report it as a valid code signing cert without it

File: script/codesign/generate-identity.sh (hunk 2 above)
Comment (MarshallOfSound) [reply script/codesign/generate-identity.sh]: The `f` scared me a bit when running locally, I can change it for CI 😆

File: script/codesign/gen-trust.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+const fingerprint = fingerprintResult.stdout.toString().replace(/^SHA1 Fingerprint=/, '').replace(/:/g, '').trim()
+
+const serialResult = cp.spawnSync('openssl', ['x509', '-serial', '-noout', '-inform', 'der', '-in', certificatePath])
+if (serialResult.status !== 0) {
+  console.error(serialResult.stderr.toString())
+  process.exit(1)
+}
+
+let serialHex = serialResult.stdout.toString().replace(/^serial=/, '').trim()
+// Pad the serial number out to 18 hex chars
```
Comment (nornagon) [inline script/codesign/gen-trust.js]: my friend, there's an npm module for this! https://www.npmjs.com/package/left-pad


...which actually has a very helpful suggestion: [String.prototype.padStart](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/padStart)

**Pull request merged**